### PR TITLE
Unify Platform interface

### PR DESCRIPTION
**Pull request by adcxyz**
Purpose and Motivation
----------------------
As proposed and favorably discussed in #4005, the Platform interface can be unified by creating class redirecting class methods for all calls to the single instance ```thisProcess.platform```. 
This PR adds all these methods, e.g. 
```
thisProcess.platform.killAll("supernova"); 
// can now be written as
Platform.killAll("supernova"); 
```
---------
- [x] All tests are passing
(No new tests written, because the PR adds no functionality, just redirections)
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by brianlheim** [comment]
I don't know how I feel about this now. It's just a lot of duplication and I thought we were going to be deprecating and removing the instance methods.

**Comment by adcxyz** [comment]
> I don't know how I feel about this now. It's just a lot of duplication and I thought we were going to be deprecating and removing the instance methods.

Well, it's less duplication (in the help files) than it was, and it is a consistent interface now. 
Deprecating all the instance methods right away seemed a bit rash to me, so I left them in, and just explain the recommended usage. Happy deprecate them now if this is agreed on - just let me know.

**Comment by jamshark70** [comment]
I tend to agree with Brian. If we're doing this, then pick a release, give people a lot of lead time, and don't stretch out the deprecation period.

It's going to hurt to remove "10 ways to do x," but I think it'll hurt less than carrying the weight for a long time.

Especially if we could provide some sed scripts to help? Like the way that LilyPond ships scripts (I think in Python) to update user code from older to newer versions (automatically making syntax changes).

**Comment by brianlheim** [comment]
Would like a comment on what i posted in #4005 before discussing this PR further.

**Comment by adcxyz** [comment]
Actually moving the implementations to the class would also save one method indirection inherent in thisProcess.platform.

**Comment by adcxyz** [comment]
@brianlheim @snappizz 
just to check whether I understand what you meant, is this it?
```
/// Move all implementations to class methods in Platform.sc, e.g.
---
	...
	*recompile {
		_Recompile
		^this.primitiveFailed
	}
	...
---
/// move all instance methods to deprecated/Platform.sc, 
/// with redirections to class method, e.g.
---
	...
	recompile {
		this.deprecated(thisMethod, Platform.findMethod(\recompile));
		^this.class.recompile
	}
...
---
```

**Comment by brianlheim** [comment]
I'm on board with this now!

File: HelpSource/Classes/Platform.schelp [hunk 1]
```diff
@@ -178,6 +137,16 @@ code::
...
 ::
 
+method:: recompile
+recompile the class library
+
+method:: clearMetadata
+clear metadata at the given path
+
+method:: devLoc
+???
```
Comment (brianlheim) [inline HelpSource/Classes/Platform.schelp]: there is usage for this, it's related to SVN which we no longer use. We should deprecate it.

File: HelpSource/Classes/Platform.schelp [hunk 2]
```diff
@@ -1,22 +1,35 @@
...
+In any running SuperCollider system, there is only a single platform object which represents the current operating system / platform. This object is accessible through the code::platform:: method of the main process instance:
 code::
 thisProcess.platform
 ::
 
-Currently implemented platforms include: OSXPlatform, LinuxPlatform, WindowsPlatform, UnixPlatform.
-
 classmethods::
-Most of Platforms class methods are simply wrappers to code::thisProcess.platform.method::.
+Note:: Almost all platform methods can be called as class methods of code::Platform::. As this is the recommended interface (for simplicity and clarity), they are documented only as class methods. So, almost all code that uses code::thisProcess.platform:: can be simplified, e.g.:
```
Comment (brianlheim) [inline HelpSource/Classes/Platform.schelp]: e.g. is unnecessary here

File: HelpSource/Classes/Platform.schelp [hunk 3]
```diff
@@ -1,22 +1,35 @@
...
+summary:: handle cross-platform differences
 categories:: Platform
 
 description::
 The Platform class (along with its subclasses) handles things which differ between operating systems (mac/linux/windows/...), to simplify cross-platform aspects of SuperCollider.
+Currently implemented platforms include: OSXPlatform, LinuxPlatform, WindowsPlatform, UnixPlatform.
 
-Platform is an abstract class encapsulating various platform dependent constants and properties, such as directories, primitive features and startup files. The platform object is accessible through the code::platform:: method of the main process instance:
+Platform is an abstract class encapsulating various platform dependent constants and properties, such as directories, primitive features and startup files.
+In any running SuperCollider system, there is only a single platform object which represents the current operating system / platform. This object is accessible through the code::platform:: method of the main process instance:
```
Comment (brianlheim) [inline HelpSource/Classes/Platform.schelp]: Probablg better to put the discussion of recommended usage up here.

File: SCClassLibrary/Platform/Platform.sc [hunk 4]
```diff
@@ -18,12 +18,18 @@ Platform {
 		recordingsDir = this.userAppSupportDir +/+ "Recordings";
 	}
 
+	// cannot redirect 'name' to name because class has a name
```
Comment (brianlheim) [inline SCClassLibrary/Platform/Platform.sc]: let's implement platformName as an instance member and deprecate name so we can make the interface consistent in the next release

**Comment by adcxyz** [comment]
@brianlheim 
> I'm on board with this now!

Great :-) 
I had one more sneaky idea how to make this entirely painless:
When all methods are class methods, and ```thisProcess.platform``` 
returns the current Platform subclass, and not the instance, 
then all should be well already: 
General implementation is in the Platform class; 
platform-specific methods are class methods in the Platform subclasses;
all user-code calls to thisProcess.platform go the subclass (instead of the instance)
which fully knows how to respond to all of them already. 
So, no need for any deprecation pain at all!

Does anyone see a reason why this would not work? 

best a

**Comment by adcxyz** [comment]
> Does anyone see a reason why this would not work?

Platform.name would need to be changed, and that would break user code that checks 
```thisProcess.platform.name``` by hand rather than using ```Platform.case```. Hmm.

**Comment by brianlheim** [comment]
I don't understand what the first comment above is asking. Regarding your last comment, I already made a note about deprecating Platform:-name in my review above.

**Comment by adcxyz** [comment]
> I don't understand what the first comment above is asking. Regarding your last comment, I already made a note about deprecating Platform:-name in my review above.

You are right, that was opaque. I am considering a stronger option, which is:
- Platform is a class with subclasses for every platform, and no instances.
- All methods are class methods
- thisProcess.platform returns the current Platform subclass, e.g. OSXPlatform
- then all user code calls to thisProcess.platform are handled by the concrete Platform subclass
- all instance methods can be removed without any need for deprecation
- only hitch: direct calls to thisProcess.platform.name will be fail, 
because XYZPlatform.name returns the class name, not the OS nametag. \linux, \osx etc.

So this option is clearer, simpler and cleaner, but an API change because of .name. 
And I was asking if anyone sees any other problems with this option.

**Comment by brianlheim** [comment]
Am I understanding correctly, that the object returned by `thisProcess.platform` in that branch is now WindowsPlatform instead of an instance of WindowsPlatform? I don't support that as a breaking change, who knows who is relying on that return type.

**Comment by brianlheim** [comment]
Please, if you want to keep discussing high level ways of approaching this, let's close this PR and discuss it on the issue ticket for the sake of everyone reading later.

**Comment by adcxyz** [comment]
closing for now, to finish discussion on the ticket #4005 first.

**Comment by adcxyz** [comment]
> Am I understanding correctly, that the object returned by thisProcess.platform in that branch is now WindowsPlatform instead of an instance of WindowsPlatform? I don't support that as a breaking change, who knows who is relying on that return type.

yes, that was the idea, and yes that is a breaking API change. 
The elegance of it is that it drops the unnecessary instance completely; 
the risk is that very low level use code may fail and need to be adapted. 
since all former instance methods except .name are supported by the concrete subclass now, 
that chance should be very low. 
So we may eventually agree on considering the benefit worth that little bit of pain. 
It's simply a more radical proposal.

**Comment by brianlheim** [comment]
Closing this PR since it's gone stale and there is still discussion needed on #4005.

**Pull request closed without merging**